client/src/components/__tests__/Signup.test.js

```diff
@@ -0,0 +1,87 @@
+import React from 'react';
+import { mount } from 'enzyme';
+
+import Signup from '../Signup';
+
+describe('<Signup/> form', () => {
+  const toggleModal = jest.fn();
+  const wrapper = mount(<Signup show toggleModal={toggleModal} />);
+
+  test('should return <FormDialog /> component', () => {
+    expect(wrapper.exists('FormDialog')).toBe(true);
+  });
+
+  test('should contain <form /> element', () => {
+    expect(wrapper.exists('form')).toBe(true);
+  });
+
+  test('form should have onsubmit method', () => {
+    expect(wrapper.find('form').props().onSubmit).toBeInstanceOf(Function);
+  });
+});
+
+describe('input fields in form', () => {
+  const toggleModal = jest.fn();
+  const wrapper = mount(<Signup show toggleModal={toggleModal} />);
+  const inputFields = wrapper.find('input');
+
+  test('should have 4 input fields', () => {
+    expect(inputFields.length).toBe(4);
+  });
+
+  test('should have email input field with onChange', () => {
+    expect(inputFields.at(0).props().name).toBe('email');
+    expect(inputFields.at(0).props().type).toBe('email');
+    expect(inputFields.at(0).props().onChange).toBeInstanceOf(Function);
+  });
+  test('should have username input field with onChange', () => {
+    expect(inputFields.at(1).props().name).toBe('username');
+    expect(inputFields.at(1).props().type).toBe('text');
+    expect(inputFields.at(1).props().onChange).toBeInstanceOf(Function);
+  });
+  test('should have password input field with onChange', () => {
+    expect(inputFields.at(2).props().name).toBe('password');
+    expect(inputFields.at(2).props().type).toBe('password');
+    expect(inputFields.at(2).props().onChange).toBeInstanceOf(Function);
+  });
+  test('should have repeat-password input field with onChange', () => {
+    expect(inputFields.at(3).props().name).toBe('repeat-password');
+    expect(inputFields.at(3).props().type).toBe('password');
+    expect(inputFields.at(3).props().onChange).toBeInstanceOf(Function);
+  });
+});
+
+describe('<Signup /> contains cancel and submit button', () => {
+  const toggleModal = jest.fn();
+  const wrapper = mount(<Signup show toggleModal={toggleModal} />);
+  const buttons = wrapper.find('button');
+
+  test('should contain two buttons', () => {
+    expect(buttons.length).toBe(2);
+  });
+
+  test('should contain cancel and submit button', () => {
+    expect(
+      buttons
+        .at(0)
+        .text()
+        .toLowerCase(),
+    ).toContain('cancel');
+
+    expect(
+      buttons
+        .at(1)
+        .text()
+        .toLowerCase(),
+    ).toContain('sign up');
+
+    expect(buttons.at(1).props().type).toBe('submit');
+  });
+
+  test('should close modal by calling toggleModal on clicking cancel button', () => {
+    const cancelBtn = wrapper.find('button').at(0);
+    cancelBtn.simulate('click');
+
+    expect(toggleModal).toHaveBeenCalledWith('showSignup');
+  });
+});
```

Comment:
where are tests for submit button? errors? success?

Reply:
About that, the whole submission and errors thing is handled by the `useSignupForm` hook. It returns a function called `handleSubmit` which when called, does all submission logic. 

Now, I haven't written unit tests for the Hooks yet.. But testing form submission and errors will be covered in that. 

Since `<SignupModal />` is a functional component, I couldn't really mock or stub `handleSubmit` function. I can't really check if `handleSubmit` function is actually called in the tests for `<Signupmodal/>`.

Reply:
Sorry I deleted my earlier comment because I got a little confused, here's my reply.

1. `<SignupModal />` uses the `<FormDialog />` component which accepts `handleSubmit` function. This component is actually responsible for triggering the `handleSubmit` function and I have tested it here: 
https://github.com/pesto-students/rattle-battle/blob/b1d288a83b0c3fb175081998affa6955c6a394d6/client/src/components/__tests__/FormDialog.test.js#L24-L32


2. The actual logic for submission and errors is handled by the `useSignupForm` hook. I haven't written unit tests for react hooks because I was still in process of figuring out how to tests for it. I'll probably make another pr adding tests for these hooks which will check all these things.

3. Since `handleSubmit` function is returned by the `useSignupForm` hook. It couldn't really mock or stub it so no tests for it in SignupModal test file. But I don't think it matters because point 1.